### ISSUE TITLE
5G-SA network not detected on Fairphone 4.

### ISSUE THREAD

**Issue by r00tb33** [herlesupreeth/docker_open5gs#243]
Hi,

I am running a SA deployment along with srsgNB but my UE i.e. fairphone 4 5G could not detect the network.

I am using a sysmocom nano SIM and have added the details of the same to database using Open5GS web-UI.

What params do I need to change in order to detect it on the UE?

Thanks!!

**Comment by herlesupreeth** [comment]
From Fairphone 4 5G webpage I see phone supporting only NSA. Can you confirm with them whether it support 5G SA?

![image](https://github.com/herlesupreeth/docker_open5gs/assets/4115839/12263fa1-f1b7-4ef2-8372-fd53b19d21ae)

**Comment by r00tb33** [comment]
> Have you tried manually searching for the 5G network?? And, can you please describe your setup?? which sdr you are using? what band are you testing? etc
> 
> Also, it could be that 5G SA i not enabled on that phone by the firmware or the SDR you are using has a drifting clock for which external clock maybe required

Yes I've tried manually searching for the network on `Fairphone 4` and `iPhone 14` using a sysmocom SIM whose details are already in the mongo DB.

My setup consists of an `intel 11th Gen i5-1135G7,  32GB RAM, 1TB SSD` with `Ubuntu 22.04.3 LTS` running on top of it. My `docker version is  24.0.6` and `docker compose version`  is `v2.21.0.`

I am using `USRP B210` which has an internal clock for GPS alternatively I also have a `bladeRF Micro A4`.  

The srsran is configured to run on `band = 78(NR)` with `dl_arfcn =  632628`.

I dunno if 5G NR is enabled on iphone 14 or not is there a way I can test this?

Thanks for the response!!

**Comment by r00tb33** [comment]
> From Fairphone 4 5G webpage I see phone supporting only NSA. Can you confirm with them whether it support 5G SA?
> 
> ![image](https://user-images.githubusercontent.com/4115839/276308805-12263fa1-f1b7-4ef2-8372-fd53b19d21ae.png)

Yes it seems Fairphone 4 5G does not support SA bands.

In that case is there a way I can run a 5G-NSA mode network?  What changes do I need to do in the current config? Or do I need to clone another image for NSA?

Please guide.

Thanks!!

**Comment by herlesupreeth** [comment]
NSA mode is not supported in open5gs so it cant be tested

**Comment by r00tb33** [comment]
Hi,

Thanks for the info!!

Also, I tried the SA deployment on iPhone 14 but could not get it detected... What could be the reason?

I've hardly made any changes to the config that came along with the docker image.

Let me know how can I debug this further.

Thanks.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by herlesupreeth** [comment]
For iPhone 14, there is a requirement to have certain 5G files in SIM card to be programmed. For the SIM you can use the following pysim script https://github.com/herlesupreeth/docker_open5gs/blob/master/sim/iphone-private-5g.script file to program your SIM card with relevant files (Dont forget to change the ADM key as per your SIM card in that script before executing). You have to execute as follows:

`./pySim-shell.py -p 0 --script scripts/iphone-private-5g.script`

In addition to the above requirement, Apple requires following as well:
1. Enable NAS encryption which can be done with following changes in amf.yaml in amf folder of this repository

    security:
        integrity_order : [ NIA2, NIA1, NIA0 ]
        ciphering_order : [ NEA2, NEA1, NEA0 ]

2. Enable user plane encryption at gNB which can be done only through code changes in srsRAN_Project gNB

In the file lib/rrc/ue/procedures/rrc_security_mode_command_procedure.cpp you would have to change as following

```
  security::preferred_ciphering_algorithms ciph_algo_pref_list = {security::ciphering_algorithm::nea2,
                                                                  security::ciphering_algorithm::nea0,
                                                                   security::ciphering_algorithm::nea1,
                                                                   security::ciphering_algorithm::nea3};
```

3. Also, you would have to use the PLMN 999 xx assigned for private 5G network

I have tested iPhone 13 Pro but the attach didn't complete due to failing after Security Mode Complete. Its a bug in srsRAN_Project gNB, for which a fix is being worked on.
